### DISCUSSION TITLE
バッチ挿入について

### DISCUSSION THREAD

**Discussion by gikeihi** [domaframework/doma#852] [Q&A] [answered]
Criteria APIのバッチ挿入について、調べるところですが、
debugすると、結局一件づつを実行して、Statement.executeBatch()を実行していないらしいです。
ソースを調べたら、
org.seasar.doma.jdbc.dialect.Dialect#supportsIdentityReservation()はfalseなので、バッチ挿入をサーポットしないです。

どうやって、Criteria APIはバッチ挿入できますか？

DB:MySQL 5,7
doma2: 2.51.0

**Comment by nakamura-to** [comment, marked as answer]
下記のようにエンティティのIDに`GenerationType.IDENTITY`を指定しているでしょうか？

```class
@Entity(metamodel = @Metamodel)
class Employee {
  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  Integer id;
}
```

この場合、本当のバッチ挿入はサポートしていません。
明示的にIDをセットしたり（MySQLではサポートされていませんが）シーケンスでIDを生成したりする場合にのみ本当のバッチ挿入が実行されます。

**Reply by gikeihi** [reply]
ご回答ありがとうございました。
確かに`GenerationType.IDENTITY`を指定しています。
ちなみに`本当のバッチ挿入はサポートしていません`というのは、doma2の仕様ですかね。
何のためですか？自動生成されたIDの値を取得するためですか？
もしバッチ挿入の場合、IDの値を要らなく、本当のバッチ挿入について、一番簡単の実現方法は何ですか？

**Reply by nakamura-to** [reply]
>自動生成されたIDの値を取得するためですか？

そうです。バッチ更新時に自動生成されたIDの取得をサポートしないデータベースがいくつかあり、Doma2はそれらのデータベースに合わせた仕様になっています。

>もしバッチ挿入の場合、IDの値を要らなく、本当のバッチ挿入について、一番簡単の実現方法は何ですか？

現状Doma2ではサポートしていないので、Doma2を使わずJDBCを直接使うのが一番簡単かもしれません。

将来のバージョンで自動生成されたIDを取得しないで本当のバッチ処理を行うオプションを導入することを検討したいと思います。

**Reply by gikeihi** [reply]
ご回答ありがとうございます。
> 将来のバージョンで自動生成されたIDを取得しないで本当のバッチ処理を行うオプションを導入することを検討したいと思います。

期待しております。

**Reply by nakamura-to** [reply]
#861 で対応します

**Reply by gikeihi** [reply]
対応していただき、ありがとうございました。
早速試します。
